sample_apps_reinvent2021/go/devops.sh

```diff
@@ -0,0 +1,227 @@
+        AND 
+measure_name = 'metrics'
+        AND time > ago(2h)
+    GROUP BY hostname
+)
+SELECT FILTER(interpolated_cpu_utilization, x -> x.value > 70.0) AS cpu_above_threshold,
+    REDUCE(FILTER(interpolated_cpu_utilization, x -> x.value > 70.0), 0, (s, x) -> s + 1, s -> s) AS count_cpu_above_threshold,
+    ROUND(REDUCE(interpolated_cpu_utilization, CAST(ROW(0, 0) AS ROW(count_high BIGINT, count_total BIGINT)),
+        (s, x) -> CAST(ROW(s.count_high + IF(x.value > 70.0, 1, 0), s.count_total + 1) AS ROW(count_high BIGINT, count_total BIGINT)),
+        s -> IF(s.count_total = 0, NULL, CAST(s.count_high AS DOUBLE) / s.count_total)), 4) AS fraction_cpu_above_threshold
+FROM time_series_view"
+ 
+#10. List the measurements with CPU utilization lower than 75% for a specific EC2 host over the past 2 hours, filling in the missing values using linear interpolation.
+go run query-sample.go -query "WITH time_series_view AS (
+    SELECT min(time) AS oldest_time, INTERPOLATE_LINEAR(
+        CREATE_TIME_SERIES(time, ROUND(cpu_utilization, 2)),
+        SEQUENCE(min(time), max(time), 10s)) AS interpolated_cpu_utilization
+    FROM $1.$2
+    WHERE 
+ hostname = '$3'
+        AND 
+measure_name = 'metrics'
+        AND time > ago(2h)
+    GROUP BY hostname
+)
+SELECT FILTER(interpolated_cpu_utilization, x -> x.value < 75 AND x.time > oldest_time + 1m)
+FROM time_series_view"
+ 
+#11. Find the total number of measurements with of CPU utilization of 0% for a specific EC2 host over the past 2 hours, filling in the missing values using linear interpolation.
+go run query-sample.go -query "WITH time_series_view AS (
+    SELECT INTERPOLATE_LINEAR(
+        CREATE_TIME_SERIES(time, ROUND(cpu_utilization, 2)),
+        SEQUENCE(min(time), max(time), 10s)) AS interpolated_cpu_utilization
+    FROM $1.$2
+     WHERE 
+ hostname = '$3'
+        AND 
+measure_name = 'metrics'
+        AND time > ago(2h)
+    GROUP BY hostname
+)
+SELECT REDUCE(interpolated_cpu_utilization,
+    DOUBLE '0.0',
+    (s, x) -> s + 1,
+    s -> s) AS count_cpu
+FROM time_series_view"
+
+#12. Find the average CPU utilization for a specific EC2 host over the past 2 hours, filling in the missing values using linear interpolation.
+go run query-sample.go -query "WITH time_series_view AS (
+    SELECT INTERPOLATE_LINEAR(
+        CREATE_TIME_SERIES(time, ROUND(cpu_utilization, 2)),
+        SEQUENCE(min(time), max(time), 10s)) AS interpolated_cpu_utilization
+    FROM $1.$2
+     WHERE 
+ hostname = '$3'
+     AND 
+measure_name = 'metrics'
+        AND time > ago(2h)
+    GROUP BY hostname
+)
+SELECT REDUCE(interpolated_cpu_utilization,
+    CAST(ROW(0.0, 0) AS ROW(sum DOUBLE, count INTEGER)),
+    (s, x) -> CAST(ROW(x.value + s.sum, s.count + 1) AS ROW(sum DOUBLE, count INTEGER)),
+     s -> IF(s.count = 0, NULL, s.sum / s.count)) AS avg_cpu
+FROM time_series_view"
+
+# Try other functionality in query
+go run query-operations.go -query "SELECT * FROM $1.$2"
```

Comment:
Can we rename the script as query-with-cancel.go? Also move out the limit query to this script.

Reply:
Its a good suggestion but lets take this refactoring later as part of upcoming changes which are planned.